walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,10 +15,157 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import java.lang.Boolean;
+import java.util.HashSet;
+import java.util.HashMap;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-    //idea use recursion over all attendees and see if you can find a start time that works for everyone
+
+  public Collection<TimeRange> calendarQuery(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    ArrayList<Event> listEvents = new ArrayList<>(events);
+    HashSet<String> mandatoryAttendees = new HashSet<>(request.getAttendees());
+    ArrayList<Event> sortedListEvents = mySort(listEvents, mandatoryAttendees);
+
+    int rangeStart = 0;
+    TimeRange previousSlot = null;
+    for (Event event: sortedListEvents) {
+      TimeRange eventTimeRange = event.getWhen();
+      // Checks for overlapping events
+      if (previousSlot == null || !eventTimeRange.overlaps(previousSlot)) {
+        availableTimes.add(TimeRange.fromStartEnd(rangeStart, eventTimeRange.start(), false));
+        rangeStart = eventTimeRange.end();
+        previousSlot = eventTimeRange;
+      } else if(previousSlot != null && previousSlot.contains(eventTimeRange)) {
+        // Checks for nested events
+        rangeStart = previousSlot.end();
+      } else {
+        rangeStart = eventTimeRange.end();
+        previousSlot = eventTimeRange;
+      }
+    }
+     
+    /* If the last event is not at the end of the day then add
+    * the timerange from the end of last event to the end of the day. */
+    if (rangeStart != 60 * 24) {
+      availableTimes.add(TimeRange.fromStartEnd(rangeStart, 60 * 24, false));
+    }
+
+    // Gets rid of short time slots
+    for (int i = 0; i < availableTimes.size(); i++ ) {
+      if (availableTimes.get(i).duration() < request.getDuration()) {
+        availableTimes.remove(i);
+        i--;
+      }
+    }
+
+    ArrayList<TimeRange> optionalTimes = new ArrayList<TimeRange>();
+    HashSet<String> optionalAttendees = new HashSet<>(request.getOptionalAttendees());
+    if (optionalAttendees.size() > 0) {
+      optionalTimes = optimizeOptionalAttendees(availableTimes, request, optionalAttendees, events);
+    }
+
+    // Returns timeranges with optional attendess if any exist 
+    return (optionalTimes.size() == 0) ? availableTimes : optionalTimes;
   }
-}
+  
+  /*
+  * Sorts events by start time and filters out events that mandatoryAttendees are not apart of.
+  */
+  public ArrayList<Event> mySort(ArrayList<Event> list, HashSet<String> mandatoryAttendees) {
```

Comment:
consider using a more descriptive, impersonal name (e.g. eventSort, for example)

Reply:
done.